tempesta_fw/t/unit/test_http_parser.c

```diff
@@ -85,45 +165,64 @@ reset_req(size_t data_len)
  *  >  0 - EOF: all possible fragments are parsed, terminate the loop.
  */
 int
-do_split_and_parse(unsigned char *req_str)
+do_split_and_parse(unsigned char *str, int type)
 {
 	static char head_buf[PAGE_SIZE];
 	static char tail_buf[PAGE_SIZE];
-	static size_t req_len, head_len, tail_len;
+	static size_t len, head_len, tail_len;
+	static int chunks;
 	int r;
 
-	BUG_ON(!req_str);
-	BUG_ON(head_len > req_len);
+	BUG_ON(!str);
+	BUG_ON(head_len > len);
 	BUG_ON(head_len > sizeof(head_buf));
 	BUG_ON(tail_len > sizeof(tail_buf));
-	reset_req(strlen(req_str));
+	if (type == FUZZ_REQ)
+		reset_req(strlen(str));
+	else if (type == FUZZ_RESP)
+		reset_resp(strlen(str));
+	else
+		BUG();
 
 	/* First iteration. */
-	if (!req_len) {
-		req_len = strlen(req_str);
+	if (!len) {
+		len = strlen(str);
 		head_len = 0;
-		tail_len = req_len - head_len;
+		tail_len = len;
+		chunks = 2;
 
-		BUG_ON(req_len > sizeof(head_buf));
-		memcpy(head_buf, req_str, req_len);
+		BUG_ON(len > sizeof(head_buf));
+		memcpy(head_buf, str, len);
 
 		/* Parse request as a single chunk on the first iteration. */
-		return tfw_http_parse_req(req, head_buf, req_len);
+		if (type == FUZZ_REQ)
+			return tfw_http_parse_req(req, head_buf, len);
+		else
+			return tfw_http_parse_resp(resp, head_buf, len);
+	}
+
+	while (chunks > 2) {
+		r = split_and_parse_n(str, type, len, chunks);
+		/* Done all iterations? */
+		if (chunks == len) {
+			len = head_len = tail_len = 0;
+			return r;
+		}
+		++chunks;
 	}
 
 	++head_len;
 	--tail_len;
 
-	/* Done all iterations?. */
-	if (head_len == req_len) {
-		req_len = head_len = tail_len = 0;
-		return 1;
+	if (head_len == len) {
+		++chunks;
+		return 0;
```

Comment:
Here we go to the testing of the request after the return success (0) without parsing.